apps/sms/services/js/conversation/conversation_client.js

```diff
@@ -0,0 +1,77 @@
+/* global bridge,
+          streamClient,
+          Thread
+*/
+
+(function(exports) {
+  'use strict';
+
+  var debug = 0 ?
+    (arg1, ...args) => console.log('[ConversationClient] ' + arg1, ...args):
+    () => {};
+
+  /**
+   * Name of the service that is responsible for managing conversation.
+   * @type {string}
+   */
+  const SERVICE_NAME = 'conversation-service';
+
+  /**
+   * Reference to active bridge client instance.
+   *
+   * @type {Client}
+   */
+  var client;
+
+  /**
+   * ConversationClient is a wrapper around bridge client connected to the
+   * bridge service hosted in a Shared Worker.
+   * @type {Object}
+   */
+  var ConversationClient = {
+    /**
+     * Initialized conversation service client bridge.
+     */
+    init() {
+      var serviceEndpoint = new SharedWorker(
+        '/services/js/conversation/conversation_service.js'
+      );
+
+      serviceEndpoint.port.start();
+
+      client = bridge.client(SERVICE_NAME, serviceEndpoint.port).plugin(
+        streamClient
+      );
+    },
+
+    /**
+     * Retrieves all conversations and drafts.
+     * @param {Function} onConversationRetrieved Callback that is called for
+     * every retrieved conversation.
+     * @returns {Promise} Promise is resolved once all conversations are
+     * retrieved and rejected if conversation stream is closed unexpectedly.
+     */
+    getAllConversations(onConversationRetrieved) {
+      var getAllConversationsStream = client.stream('getAllConversations');
+
+      getAllConversationsStream.listen((thread) => {
+        debug(
+          'Conversation with id %s is retrieved from stream', thread.id
+        );
+
+        onConversationRetrieved(new Thread(thread));
```

Comment:
> my concern here is that we're calling onConversationRetrieved synchronously; might be one of the reasons of the slowness ?

Hmm, I'll check, but I was thinking that this should not affect performance overall. I mean that this doesn't stop cursor from iterating, once thread is retrieved, it's passed further via asynchronous postMessage from iframe to ShW  (via BroadcastChannel) and then again asynchronously through MessageChannel to Window. I was assuming we render thread faster then next thread goes through this chain.... But as always need to measure first :)